app/services/exchange_rates/average_exchange_rates_service.rb

```diff
@@ -0,0 +1,97 @@
+module ExchangeRates
+  class AverageExchangeRatesService
+    VALID_MONTHS = [4, 12].freeze
+    VALID_DAY = 31
+
+    def self.call(force_run:, selected_date:)
+      new(force_run, selected_date).call
+    end
+
+    def initialize(force_run, selected_date)
+      @force_run = force_run
+      @selected_date = Date.parse(selected_date)
+    end
+
+    def call
+      return argument_error unless valid_date || force_run
+
+      avg_rates = create_average_rates
+
+      countries_and_rate = match_country_with_rates(avg_rates)
+      upload_average_rate_file(countries_and_rate)
+    end
+
+    private
+
+    attr_reader :force_run, :selected_date
+
+    def match_country_with_rates(avg_rates)
+      live_countries_last_twelve_months.each_with_object({}) do |country, h|
+        # Need to ensure unique rates are being collected
+        avg_rate = avg_rates
+                    .select { |rate| rate.currency_code == country.currency_code }
+                    .first
+                    .rate
+
+        # { ExchangeRateCountryCurrency => avg_rate }
+        h[country] = avg_rate
+      end
+    end
+
+    def upload_average_rate_file(countries_and_rate)
+      ExchangeRates::UploadFileService.new(
+        countries_and_rate,
+        selected_date,
+        :average_csv,
+      ).call
+    end
+
+    def valid_date
+      return false unless selected_date.day == VALID_DAY
+      return false unless VALID_MONTHS.include?(selected_date.month)
+
+      true
+    end
+
+    def create_average_rates
+      live_countries_last_twelve_months.select_map(:currency_code).uniq.map do |currency_code|
+        avg_rate = valid_average_rate(currency_code)
+
+        next unless avg_rate
```

Comment:
We should avoid using a next inside a map.
I'd suggest to use reject block before, to filter out the elements you want to exclude. 
```
something.reject{ |currency_code| valid_average_rate(currency_code).nil?}
```

Reply:
will think about this one

Reply:
I changed the logic in the end, reject wouldnt work as thats on the a different element so hopefully it works better now